JustSaying/Messaging/MessageSerialisation/MessageSerialisationRegister.cs

```diff
@@ -1,57 +1,71 @@
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using JustSaying.Models;
 
 namespace JustSaying.Messaging.MessageSerialisation
 {
     public class MessageSerialisationRegister : IMessageSerialisationRegister
     {
+        // TODO Inconsistent use of American and British English for "serialize"/"serialise"
+
         private readonly IMessageSubjectProvider _messageSubjectProvider;
-        private readonly Dictionary<Type, TypeSerialiser> _map = new Dictionary<Type, TypeSerialiser>();
+        private readonly IDictionary<Type, TypeSerialiser> _map = new ConcurrentDictionary<Type, TypeSerialiser>();
 
         public MessageSerialisationRegister(IMessageSubjectProvider messageSubjectProvider)
         {
-            _messageSubjectProvider = messageSubjectProvider;
+            _messageSubjectProvider = messageSubjectProvider ?? throw new ArgumentNullException(nameof(messageSubjectProvider));
         }
 
         public void AddSerialiser<T>(IMessageSerialiser serialiser) where T : Message
         {
             var key = typeof(T);
-            if (!_map.ContainsKey(key))
+            if (!_map.TryGetValue(key, out TypeSerialiser serializer))
             {
-                _map.Add(key, new TypeSerialiser(typeof(T), serialiser));
+                _map[key] = new TypeSerialiser(typeof(T), serialiser);
```

Comment:
Since _map_ is private, there is no reason not to type it as `ConcurrentDictionary<>`  which would allow you to use the atomic methods `TryAdd` / `GetOrAdd` / `AddOrUpdate ` here.

As it stands, there's not much difference between this and the original code in terms of concurrency.

Reply:
`ConcurrentDictionary<K, V>` conforms to `IDictionary<K, V>` under the hood when used as such, so you can use it "simpler" while not getting the possibility of a `NullReferenceException` if two adds collide and corrupt the internal state, which is what happens to `Dictionary<K, V>`.

Reply:
So `ConcurrentDictionary`'s implementation of `_map[key] = value;` is resistant to concurrency issues that you can get with `Dictionary` ? Cool, I did not know that.

Reply:
This has been one of those questions I've wondered, but never taken the time to read the docs on. TIL 👍